sklearn/hpsearch/gke_parallel.py

```diff
@@ -0,0 +1,437 @@
+            self.output_uris[worker_id] = 'gs://{}/{}/{}/fitted_search.pkl'.format(self.bucket_name, self.task_name, worker_id)
+            self.output_without_estimator_uris[worker_id] = 'gs://{}/{}/{}/fitted_search_without_estimator.pkl'.format(self.bucket_name, self.task_name, worker_id)
+            self.dones[worker_id] = False
+
+            pickle_and_upload(search_spaces, self.bucket_name, '{}/{}/search_spaces.pkl'.format(self.task_name, worker_id))
+
+            self._deploy_job(worker_id, X_uri, y_uri)
+
+
+    def _upload_data(self, X, y):
+        if type(X) == str and X.startswith('gs://'):
+            X_uri = X
+        else:
+            X_uri = pickle_and_upload(X, self.bucket_name, '{}/X.pkl'.format(self.task_name))
+
+        if type(y) == str and y.startswith('gs://'):
+            y_uri = y
+        else:
+            y_uri = pickle_and_upload(y, self.bucket_name, '{}/y.pkl'.format(self.task_name))
+
+        search_uri = pickle_and_upload(self.search, self.bucket_name, '{}/search.pkl'.format(self.task_name))
+
+        return X_uri, y_uri, search_uri
+
+
+    def fit(self, X, y):
+        """Returns an `operation` object that implements `done()` and `result()`.
+        """
+        timestamp = str(int(time.time()))
+        self.task_name = self.task_name or '{}.{}.{}'.format(self.cluster_id, self.image_name, timestamp)
+        self._done = False
+        self._cancelled = False
+
+        X_uri, y_uri, _ = self._upload_data(X, y)
+
+        if type(self.search) == GridSearchCV:
+            handler = self._handle_grid_search
+        elif type(self.search) == RandomizedSearchCV:
+            handler = self._handle_randomized_search
+        elif type(self.search) == BayesSearchCV:
+            handler = self._handle_bayes_search
+
+        print('Fitting {}'.format(type(self.search)))
+        handler(X_uri, y_uri)
+
+        self.persist()
+
+
+    def persist(self):
+        self.gcs_uri = pickle_and_upload(self, self.bucket_name, '{}/gke_search.pkl'.format(self.task_name))
+        print('Persisted the GKEParallel instance: {}'.format(self.gcs_uri))
+
+
+    # Implement part of the concurrent.future.Future interface.
+    def done(self):
+        # TODO: consider using kubernetes API to check if pod completed
+        if not self._done:
+            for worker_id, output_uri in self.output_uris.items():
+                print('Checking if worker {} is done'.format(worker_id))
+                self.dones[worker_id] = get_uri_blob(output_uri).exists()
+
+            self._done = all(self.dones.values())
+
+        return self._done
+
+
+    def cancel(self):
+        """Deletes the kubernetes jobs, but completed data remains."""
+        if not self._cancelled:
+            delete_jobs_pods(self.job_names.values())
+            self._cancelled = True
+
+
+    def cancelled(self):
+        return self._cancelled
+
+
+    # TODO: allow getting only the best result to save time
```

Comment:
Same as above TODO, still needed?

Reply:
Not implemented yet, might still be worth doing at some point.